search.py

```diff
@@ -406,101 +406,92 @@ def astar_search(problem, h=None):
-        elif action=='LEFT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)-1]
-            new_state[blank_square(0)][blank_square(1)-1] = 0
-        elif action=='DOWN':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)+1][blank_square(1)]
-            new_state[blank_square(0)+1][blank_square(1)] = 0
-        elif action=='RIGHT':
-            new_state[blank_square(0)][blank_square(1)] = new_state[blank_square(0)][blank_square(1)+1]
-            new_state[blank_square(0)][blank_square(1)+1] = 0
+        ''' Given state and action, return a new state that is the result of the action.
+        Action is assumed to be a valid action in the state '''
+
+        # ix is the index of the blank square
+        ix = self.find_blank_square(state)
+        new_state = list(state)
+
+        if action == 'UP':
+            new_state[ix], new_state[ix - 3] = new_state[ix - 3], 0
+        elif action == 'DOWN':
+            new_state[ix], new_state[ix + 3] = new_state[ix + 3], 0
+        elif action == 'LEFT':
+            new_state[ix], new_state[ix - 1] = new_state[ix - 1], 0
+        elif action == 'RIGHT':
+            new_state[ix], new_state[ix + 1] = new_state[ix + 1], 0
         else:
-            print("Invalid Action!")
-        return new_state
+            print('Invalid Action')
+
+        return tuple(new_state)
 
     def goal_test(self, state):
-        """Given a state, return True if state is a goal state or False, otherwise"""
-        for row in len(state):
-            for column in len(row):
-                if state[row][col] != self.goal[row][column]:
-                    return False
-        return True
-
-    def checkSolvability(self, state):
+        ''' Given a state, return True if state is a goal state or False, otherwise '''
+
+        if tuple(state) == self.goal:
+            return True
+        return False
+
+    def check_solvability(self, state):
+        ''' Checks if the given state is solvable '''
+
         inversion = 0
         for i in range(len(state)):
-               for j in range(i, len(state)):
-                    if (state[i] > state[j] and state[j] != 0):
-                                    inversion += 1
+            for j in range(i, len(state)):
+                if (state[i] > state[j] and state[j] != 0):
+                    inversion += 1
         check = True
-        if inversion%2 != 0:
-                check = False
-        print(check)
+
+        if inversion % 2 != 0:
+            check = False
+        return check
     
-    def h(self, state):
-        """Return the heuristic value for a given state. Heuristic function used is 
-        h(n) = number of misplaced tiles."""
+    def h(self, node):
+        ''' Return the heuristic value for a given state. Default heuristic function used is 
+        h(n) = number of misplaced tiles '''
+
```

Comment:
I would do these 5 lines plus 3 blank lines as 1 line:

      return sum(s != g for (s, g) in zip(node.state, self.goal))